File: Embassat/View/Artists/CADEMArtistsViewModel.m
//
//  CADEMArtistsViewModel.m
//  Embassat
//
//  Created by Joan Romano on 08/03/14.
//  Copyright (c) 2014 Crows And Dogs. All rights reserved.
//

#import "CADEMArtistsViewModel.h"

#import <NSValueTransformer+MTLPredefinedTransformerAdditions.h>

#import "CADEMArtistDetailViewModel.h"

static NSString *const kArtistsCacheKey = @"ArtistsCacheKey";

@interface CADEMArtistsViewModel ()

@property (nonatomic, strong) id model;

@end

@implementation CADEMArtistsViewModel

#pragma mark - Lifecycle

- (instancetype)init
{
	return [self initWithModel:nil];
}

- (instancetype)initWithModel:(id)model
{
	if (self = [super init])
    {
        _model = model;
        
        _updatedContentSignal = [RACObserve(self, model) ignore:nil];
        
        RAC(self, model) = [self.didBecomeActiveSignal
                             flattenMap:^RACStream *(id value) {
                                 return [self artists];
                             }];
    }
    
	return self;
}

#pragma mark - Public

- (NSInteger)numberOfItemsInSection:(NSInteger)section
{
    return [self.model count];
}

- (id)artistViewModelForIndexPath:(NSIndexPath *)indexPath
{
    return [[CADEMArtistDetailViewModel alloc] initWithModel:self.model[indexPath.row]];
}

- (NSString *)titleAtIndexPath:(NSIndexPath *)indexPath
{
    return [[self.model[indexPath.row] name] uppercaseString];
}

#pragma mark - Private

- (id)model
{
    return [_model sortedArrayUsingComparator:^NSComparisonResult(CADEMArtistSwift *artist1, CADEMArtistSwift *artist2) {
        return [artist1.name caseInsensitiveCompare:artist2.name];
    }];
}

- (RACSignal *)artists
{
    return [[RACSignal createSignal:^RACDisposable *(id<RACSubscriber> subscriber) {
        CADArtistService *service = [[CADArtistService alloc] init];
        [service artists:^void(NSError *error){
            [subscriber sendError:error];
        } success:^void(id artists){
            [subscriber sendNext:artists];
            [subscriber sendCompleted];
        }];
        
        return nil;
    }] replayLazily];
}

@end

File: Embassat/View/Artists/CADEMArtistsViewModel.swift
//
//  CADEMArtistsViewModel.swift
//  Embassa't
//
//  Created by Joan Romano on 17/05/15.
//  Copyright (c) 2015 Crows And Dogs. All rights reserved.
//

import Foundation
import ReactiveCocoa

public class CADEMArtistsViewModel: NSObject, CADEMViewModelCollectionDelegate {
    
    var model: Array<CADEMArtist> = []
    public let service: CADEMArtistService = CADEMArtistService()
    public let activeSubject: RACSubject
    
    override init() {
        activeSubject = RACSubject()
        
        super.init()
        
        service.artists().map
            { (artists: AnyObject!) -> AnyObject! in
                let artistsArray = artists as! Array<CADEMArtist>
                return artistsArray.sort({ (artist1, artist2) -> Bool in
                    artist1.name < artist2.name
                })
            }.subscribeNext(
                { [unowned self] (artists: AnyObject!) -> Void in
                    self.model = artists as! Array<CADEMArtist>
                    self.activeSubject.sendNext(true)
            }, error:
                { [unowned self] (error: NSError!) -> Void in
                    self.activeSubject.sendError(error)
            })
    }
    
    func numberOfItemsInSection(section : Int) -> Int {
        return self.model.count
    }
    
    public func shouldRefreshModel() {
        service.cachedArtists().map{ (artists: AnyObject!) -> AnyObject! in
            let artistsArray = artists as! Array<CADEMArtist>
            return artistsArray.sort({ (artist1, artist2) -> Bool in
                artist1.name < artist2.name
            })
        }.subscribeNext { [unowned self] (artists: AnyObject!) -> Void in
            self.model = artists as! Array<CADEMArtist>
        }
    }
    
    public func titleAtIndexPath(indexPath: NSIndexPath) -> String {
        return self.model[indexPath.row].name.uppercaseString
    }
    
    public func artistViewModel(forIndexPath indexPath: NSIndexPath) -> CADEMArtistDetailViewModel {
        return CADEMArtistDetailViewModel(model: model, currentIndex: indexPath.item)
    }
}
